src/compiler/checker.ts

```diff
@@ -6588,7 +6588,13 @@ namespace ts {
                 // in the process of resolving (see issue #6072). The temporarily empty signature list
                 // will never be observed because a qualified name can't reference signatures.
                 if (symbol.flags & (SymbolFlags.Function | SymbolFlags.Method)) {
-                    type.callSignatures = getSignaturesOfSymbol(symbol);
+                    // If the function is explicitly marked with `@class`, then it must be constructed.
+                    if (isInJavaScriptFile(symbol.valueDeclaration) && getJSDocClassTag(symbol.valueDeclaration)) {
+                        type.constructSignatures = getSignaturesOfSymbol(symbol);

```

Comment:
given the way this breaks intellisense, perhaps a better change would be to change `resolveCallExpression` to not resolve call signatures on functions with `@constructor` instead of switching them entirely to be construct signatures. Something like (just before the last `return` of the function):

```ts
            if (callSignatures.some(sig => isJavaScriptConstructor(sig.declaration) && !!getJSDocClassTag(sig.declaration!))) {
                error(node, Diagnostics.Value_of_type_0_is_not_callable_Did_you_mean_to_include_new, typeToString(funcType));
                return resolveErrorCall(node);
            }
```

Although there's probably a more elegant way to write it. This undoes the questionable baseline changes that happen with the current code.